Fix sRGB GLSL pragmas

johnnovak opened the pull request:
This is embarrassing... It doesn't even warrant the "regression" label because it couldn't have ever worked properly! :lolsob:

I guess I never noticed because I started using CRT shaders that do their own gamma decoding/encoding "manually" and thus interpolate in a  gamma-correct fashion.

Anyway, it should be good now, and it can be verified by uncommenting the debug logging. These two flags are enabled in the default `interpolation/sharp` shader and disabled in most CRT shaders, such as `crt/hyllian-updated`.

FeralChild64 commented:
After the change the video output started looking strange with my configuration - font now looks thicker, and by a lot.

```init
[sdl]
windowresolution    = 1280x960
output              = opengl
texture_renderer    = auto

[dosbox]
force_vga_single_scan       = false
vga_8dot_font               = true

[render]
aspect             = true
integer_scaling    = off
glshader           = 
```
Before:

![old](https://github.com/dosbox-staging/dosbox-staging/assets/48332137/e92c96bc-414b-4923-b8bb-b1d0d3fe3560)

After:

![new](https://github.com/dosbox-staging/dosbox-staging/assets/48332137/89ce6e65-dc0a-49c4-8e1e-75f9c146463c)

FeralChild64 commented:
Windows 3.1 also works better on main (zoom to 100% size to see the difference):

![old-win](https://github.com/dosbox-staging/dosbox-staging/assets/48332137/cec4bb0b-efa2-4828-94b1-91566208fe35)

than after the patch:

![new-win](https://github.com/dosbox-staging/dosbox-staging/assets/48332137/290027d9-5dac-4b18-8375-684e2b3f66fa)

weirddan455 commented on src/gui/sdlmain.cpp:
```diff
@@ -2165,19 +2165,30 @@ Bitu GFX_SetSize(int width, int height, const Bitu flags, double scalex,
 		                             ? GL_SRGB8_ALPHA8
 		                             : GL_RGB8;
+
+		// if (texformat == GL_SRGB8_ALPHA8) {
+		// 	LOG_MSG("OPENGL: Using sRGB texture");
+		// }
```
I've never looked at the render code in DosBox so I'm probably not the best person for review.  But I don't like commented out code like this.  I'd say either wrap this in an `#ifdef` for debug builds or just remove it.

johnnovak replied:
They definitely won't be removed as they are super useful for focused debugging sessions. Writing them over and over again is a pain. Also , I don't like using "debug" log levels because then you get flooded with all sorts of debug messages you don't care about when enabling debug (or "trace") logging.

Another non-solution is logging categories -- too complicated to maintain, and at the end of the day, often I only want one of two specific debug messages and that's it...

What could be done, perhaps, is surrounding them with `#if 0` / `#endif`, so you can toggle them on by changing 0 to 1. But... dunno, I don't have problems with the current comment based solution. It does the job with minimal hassle.

@FeralChild64 @shermp @kklobe your views?

johnnovak replied:
Btw, these were trivial one-liners. What I wrote is more about "heavy-duty" logging like this:

https://github.com/dosbox-staging/dosbox-staging/blob/main/src/capture/image/image_scaler.cpp#L139-L189

(Although, maybe a bit of a non-example as I wrapped this in a function, so I can easily only toggle calling this function... 😄)

weirddan455 replied:
Just an idea that I've seen SDL code do is wrap these in (for example) `#ifdef DEBUG_RENDERER `

Then when you're working on this specific part of the code, you can just type `#define DEBUG_RENDERER` at the top of the file (or use the build system to pass the flag in) and you get everything turned on.

But if you just want to stick to commented out code, I won't make too big of a fuss over it.  I just think it looks kind of ugly.

johnnovak replied:
I'm personally happy to switch to `#if 0` highlighting; like I said, I want to be able to toggle comments on an individual basis. Toggling them wholesale with `DEBUG_RENDERER` won't cut it as I explained. Too noisy, they most often need to be toggled individually in most debugging tasks.

So for example this

<img width="699" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/1d554b22-d6c1-404a-b30d-c1226cf57f39">

Would become this

<img width="687" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/f17121d4-5c09-484c-a161-2591726664eb">

So unless objections from others, I'm happy to start using this style going forward. I was concerned that I won't see these blocks highlighted as comments, but the Vim C++ syntax plugin does support `#if 0`/`#if 1` highlighting out of the box, so that's great 😄

johnnovak replied:
And as I said, for multi-line debug log statements changing `0` to `1` and back is actually a lot easier than toggling both ends of the comment block, or per line (if you editor doesn't support commenting/uncommenting whole blocks with one shortcut), so it was a good suggestion @weirddan455 👍🏻

weirddan455 commented on src/gui/sdlmain.cpp:
```diff
@@ -2191,6 +2202,7 @@ Bitu GFX_SetSize(int width, int height, const Bitu flags, double scalex,
 
 		if (sdl.opengl.framebuffer_is_srgb_encoded) {
 			glEnable(GL_FRAMEBUFFER_SRGB);
+		//	LOG_MSG("OPENGL: Using sRGB framebuffer");
```
More commented out code here.

johnnovak commented:
> After the change the video output started looking strange with my configuration - font now looks thicker, and by a lot.

Thanks for reporting this, something clearly went wrong there. Can't reproduce it on my Mac, trying Windows now. It's a bit perpexling because you've explicitly disabled shaders, and this fix should only make a difference with shaders enabled (and not even all shaders, just certain ones that "opt-in" to the use of sRGB textures & framebuffer via the pragmas).

Investigating...

johnnovak commented:
Good news @FeralChild64, could reproduce on my Nvidia card on Windows!

johnnovak commented:
Okay, I managed to de-confuse myself and now I understand what's going on. This is happening on all platforms, I was confused when I said this cannot be reproduced on macOS...

1. The "no shader" path with `output = opengl` and `glshader = ` (explicitly set to empty string) actually uses the `none.glsl` shader. We _always_ use a shader with OpenGL, no matter what (we have to; this is modern OpenGL); and this is the so-called "no shader" path.

2. When I introduced gamma-correct interpolation last year, I've updated all shaders with the `use_*` pragmas to ensure gamma-correct bilinear interpolation. It's just that feature never worked 😩 so you never saw the gamma-correct results with the `none` shader that you dislike, so you couldn't complain about it back then 😆 

3. My view: both gamma-correct and gamma-incorrect bilinear interpolation look like shit—but it's a slightly different shit😎 You got used to _one_ particular brand of crappiness, but in the "extreme edge cases" of white-text-on-black-background, or black-text-on-white-background, the letters look either thicker or thinner, depending on whether correct or incorrect interpolation is used. It's easier to understand this by looking at the zoom-in image example below. The TL;DR is, you can't "win" with bilinear interpolation; either black-on-white or white-on-black will look slightly off, depending on the interpolation method. Solution: use the default "sharp" shader, or better yet, use a nice CRT shader.

(As a side-note, this kinda explains why you were insistent on saying that forcing 8-dot VGA fonts looked more authentic to you. Bright grey or white characters on a black background definitely look _thinner_ than they should with gamma-incorrect bilinear interpolation you have been using for a long time, presumably, no doubt about that. So by forcing 8-dot VGA fonts, you make the vertical lines a bit thicker compared to the more standard 9-dot VGA fonts, so you're effectively adding a "fudge factor" to "bring back" the correct-appearing width of the vertical lines. Or you indeed had some really early VGA card with 8-dot fonts, that's the other explanation.)

# Zoomed-in examples

## Default `interpolation/sharp` shader (gamma-correct)

This is the "idealised" look when we treat as little sharp rectangles.

<img width="787" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/ae3a010b-399c-42e8-9312-dfde26c66983">

## `none` shader (gamma-correct bilinear interpolation)

Mathematically correct; white-on-black looks a bit thick, black-on-white looks a bit thin.

<img width="744" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/5bcf4650-5c65-42cd-9852-7c8bb0522f6d">

## `none` shader (gamma-incorrect bilinear interpolation)

Mathematically incorrect; basically some random nonsensical fudging of the numbers because that's what gamma-incorrect interpolation is (sadly this has been the default for GPUs for a long time because it used to be computationally less expensive before hardware-assisted sRGB handling via lookups tables became standard). White-on-black looks a bit thing, black-on-white looks quite thick.

<img width="762" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/e1ddc0a7-ba53-463c-8957-f541234409c3">

## CRT shader

A custom-tweaked version of `crt/hyllian-updated` I'm working on, just for reference how a good CRT shader fixes these problems. That ghastly bilinear interpolation blur is gone and the letters have optimal width in both edge cases, and the results are very similar to what you get on an actual CRT (I know because I keep comparing the results of my tweaked shader to my real CRTs).

<img width="759" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/3e873ae4-c830-4ac6-b204-565c9601675b">

# Solution

Well, not quite sure... Neither gamma-correct and gamma-incorrect straight bilinear interpolation look good at all, they're both crap in their unique and different ways 😎 Maybe we can just revert to the old "incorrect" way as that would make @FeralChild64 happy and all those people who for some reason like that look...

However, gamma-correctness is absolutely the one true and correct way for the "sharp" shader, as explained in my original ticket with numerous examples, that's non-negotiable. See that ticket for details: https://github.com/dosbox-staging/dosbox-staging/pull/1627

The new image capture feature also uses the same gamma-correct upscaling method as the "sharp" shader I implemented in software.

Gamma correctness is one of my pet peeves, and it's no exaggeration to say I'm probably a bit of an expert on the topic. The article I wrote about gamma was quite popular at HackerNews, so if you're even just a bit interested in computer graphics, then you might wanna to check it out. And if you're doing _any_ sort of image processing and you're just a bit unsure about gamma-correctness, I would say it's a _mandatory_ read.

https://news.ycombinator.com/item?id=12552094
https://blog.johnnovak.net/2016/09/21/what-every-coder-should-know-about-gamma/

---

# Full image examples

Just for reference; open them in multiple browser tabs and switch between them.

## Start screen

### `interpolation/sharp` (gamma-correct)

<img width="1445" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/940b71bc-474e-4dbd-8698-39ca7afafac6">

### `none` (gamma-correct, current PR behaviour)

<img width="1445" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/72efef95-be1b-4e63-b2e6-ec5e68b945d1">

### `none` (random gamma-incorrect result by fluke, `main` behaviour)

<img width="1445" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/68765426-3173-4f22-b315-a57e241aa936">


## Spellcasting 201

### `interpolation/sharp` (gamma-correct)

<img width="1602" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/78c1822b-07e7-4bc8-a53a-329d8d36f4f3">

### `none` (gamma-correct, current PR behaviour)

![image](https://github.com/dosbox-staging/dosbox-staging/assets/698770/6a977b03-8230-41eb-80e0-d1ef3d17f731)

### `none` (random gamma-incorrect result by fluke, `main` behaviour)

![image](https://github.com/dosbox-staging/dosbox-staging/assets/698770/8f79e21e-bd11-47e6-a4ae-254c599c0a8f)

johnnovak commented:
Can you re-test @FeralChild64 that we're back to the previous behaviour? Please make sure you're using the updated version of `none.glsl` with the `use_srgb_*` pragmas removed.

FeralChild64 commented:
Works nicely now, great!

GranMinigun commented:
Nitpick: commented-out pragmas are still parsed. And I would've preferred to leave default bilinear interpolation in the correct colour space for rare cases when I may use it, but oh well. Maybe I'll just port a couple more cubic shaders instead.

johnnovak commented:
> Nitpick: commented-out pragmas are still parsed.

Ah, ok. I'll try to shoehorn comment handling into that the regex. Did I say I hate regexes?

> And I would've preferred to leave default bilinear interpolation in the correct colour space for rare cases when I may use it, but oh well.

I may have a _slight_ preference toward the "correct" bilinear interpolation by default myself too, but I don't care much... 
As I see it, it's there mainly for people used to how bilineear with gamma-incorrect interpolation looks, otherwise everybody should use proper CRT shaders or even the default sharp shader.

GranMinigun commented:
Eh, can always drop a custom shader anyway. No biggie.
